### PR TITLE
unique_ptr Homework

### PR DESCRIPTION
**Pull request by kari0036**

### CODE REVIEW

**Review by jakjan95 — COMMENTED** [review]
Code looks good, just remember about copy and move assignments :)

File: homework/unique_ptr/unique_ptr.hpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    ptr_ = newPtr;
+}
+
+
+template <typename T>
+unique_ptr<T>::~unique_ptr() {
+    if (ptr_) {
+        delete ptr_;
+    }
+}
```
[the remark this answers is not in the document]
Comment (MaciejSalamonski) [reply homework/unique_ptr/unique_ptr.hpp:53]: I agree with @jakjan95  https://stackoverflow.com/questions/3116980/do-i-have-to-set-pointer-to-nullptr-in-destructor

**Review by MaciejSalamonski — APPROVED** [review]
Hi! Nice code! I aporove this pr 👍

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+ 
+name: CMake CI
+
+on: [push, pull_request]
+
+jobs:
+  ctest:
```
Comment (ziobron) [inline .github/workflows/main.yml:7]: add a job with valgrind / asan to prove that there are no memory leaks

File: homework/unique_ptr/unique_ptr-ut.cpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        WHEN("UseArrowOperatorToChangeMemberValue"){
+            ptr->value = secondV;
+            THEN("ShouldChangeMemberValue"){
+                CHECK(ptr->value == secondV);
+            }
+        }
+    }
+
+}
+
```
Comment (ziobron) [inline homework/unique_ptr/unique_ptr-ut.cpp:102]: good tests. There is only one thing missing - moving into a pointer that already manages another object

File: homework/unique_ptr/unique_ptr.hpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    T& operator*() const;
+    T* operator->() const { return ptr_; }
+
+private:
+    T* ptr_ = nullptr;
+};
+
+template <typename T>
+unique_ptr<T>::unique_ptr(unique_ptr&& otherPtr) {
+    ptr_ = otherPtr.release();
```
Comment (ziobron) [inline homework/unique_ptr/unique_ptr.hpp:29]: move it into constructor initializer list

File: homework/unique_ptr/unique_ptr.hpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+template <typename T>
+T* unique_ptr<T>::release() {
+    T* temp = ptr_;
+    ptr_ = nullptr;
+    return temp;
+}
+
+template <typename T>
+void unique_ptr<T>::reset(T* newPtr) {
+    if (ptr_) {
```
Comment (ziobron) [inline homework/unique_ptr/unique_ptr.hpp:41]: if not needed. deleting a nullptr is safe.

File: homework/unique_ptr/unique_ptr.hpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+template <typename T>
+class unique_ptr {
+public:
+    unique_ptr() = default;
+    unique_ptr(T* ptr)
+        : ptr_(ptr) {}
+    unique_ptr(const unique_ptr& otherPtr) = delete;
+    unique_ptr(unique_ptr&& otherPtr);
+
```
Comment (ziobron) [inline homework/unique_ptr/unique_ptr.hpp:13]: you are missing move and copy assignment operators

**Review by ziobron — CHANGES_REQUESTED** [review]

**Review by ziobron — CHANGES_REQUESTED** [review]
Looks fine, but CI tells that:
 fatal error: catch.hpp: No such file or directory
29
    2 | #include "catch.hpp"

**Review by WBigaj — COMMENTED** [review]
I have checked your code and have nothing to add that have not been mentioned by other reviewers.

**Review by ziobron — CHANGES_REQUESTED** [review]
Another CI error
valgrind: ./uniquePtr-ut: No such file or directory
##[error]Process completed with exit code 127.

**Comment by ziobron** [comment]
{ "requirement_ids": [3784, 1119] }
36 XP za implementację unique_ptra (3784)
3 XP za punktualność

**Pull request closed without merging**